collector/tcpstat.go

```diff
@@ -0,0 +1,151 @@
+// +build !notcpstat
+
+package collector
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	procTcpStat  = "/proc/net/tcp"
+	procTcp6Stat = "/proc/net/tcp6"
+)
+
+type TcpConnectionState int
+
+const (
+	TCP_ESTABLISHED TcpConnectionState = iota + 1
+	TCP_SYN_SENT
+	TCP_SYN_RECV
+	TCP_FIN_WAIT1
+	TCP_FIN_WAIT2
+	TCP_TIME_WAIT
+	TCP_CLOSE
+	TCP_CLOSE_WAIT
+	TCP_LAST_ACK
+	TCP_LISTEN
+	TCP_CLOSING
+)
+
+type tcpStatCollector struct {
+	config Config
+	metric *prometheus.GaugeVec
+}
+
+func init() {
+	Factories["tcpstat"] = NewTcpStatCollector
+}
+
+// NewTcpStatCollector takes a config struct and returns
+// a new Collector exposing network stats.
+func NewTcpStatCollector(config Config) (Collector, error) {
+	return &tcpStatCollector{
+		config: config,
+		metric: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Namespace: Namespace,
+				Name:      "tcp_connection_states",
+				Help:      "Number of connection states.",
+			},
+			[]string{"state"},
+		),
+	}, nil
+}
+
+func (c *tcpStatCollector) Update(ch chan<- prometheus.Metric) (err error) {
+	tcpStats, err := getTcpStats(procTcpStat)
+	if err != nil {
+		return fmt.Errorf("couldn't get tcpstats: %s", err)
+	}
+
+	// if enabled ipv6 system
```

Comment:
I can imagine uses for it if say you were debugging an issue with a buggy v6 client/server. I'm not too bothered either way, I think I'd tend to keep it simple for now.